lun: skip defining the lun when image not found

lxbsz opened the pull request.

idryomov commented on ceph_iscsi_config/lun.py:
```diff
@@ -778,10 +793,31 @@ def allocate(self, keep_dev_in_lio=True, in_wwn=None):
+                disk_attr = {"wwn": wwn,
+                             "image": self.image,
+                             "pool": self.pool,
+                             "datapool": self.datapool,
```
I think this "housekeeping" code needs to be extracted into a separate method.

idryomov commented on ceph_iscsi_config/lun.py:
```diff
@@ -605,6 +606,7 @@ def allocate(self, keep_dev_in_lio=True, in_wwn=None):
 
                 if not rbd_image.error:
                     self.config.add_item('disks', self.config_key)
+                    need_del_item = True
```
What is the point of adding a partially-formed config item before the operation actually succeeds?  Instead of attempting to clean up on all error paths except one, would delaying the addition until it is known that the config item is actually needed work?

idryomov commented:
Could you explain the purpose of "common: add sub item support for update_item" commit?  What is the problem are you trying to solve there?

And regardless of the reasoning behind it, there are about 40 existing `update_item` calls.  Instead of touching almost all of the them to pass an extra `None` argument, I'd introduce a new `update_sub_item` method and use it in that one or two places where you are passing in something other than `None`.

lxbsz commented:
> I see you introduced `update_sub_item`, but it's not called by anything other than `update_item` (which passes `None` for `item_name`). To me this means that either something is broken or all `update_item`-related changes are bogus.
>

Sorry, I pushed this from wrong node. Updated it.
 
> Continuing that thought, you haven't explained the purpose of "common: add sub item support for update_item" commit. What it does, why it's needed, is it a bug fix or an optimization, etc? If it's just an optimization/cleanup, let's move it to a separate PR to not stall the "image not found" bug fixes.

I just added more comments in the commit comments and code comments.

When recovery the config from the rbd image and LIO device, because the image size may changed, for example if old size is 1M and new size is 2M, then we need to sync this changes to other gateways.

Actually the first commit could fix the bz. The last 3 are one further optimization. I am okay to move them into a separate PR.

lxbsz commented:
> Continuing that thought, you haven't explained the purpose of "common: add sub item support for update_item" commit.

For this, because only the auth gateway could change the relevant `config["disks"][disk_key]`, and for all the other gateways it's readonly. Since the rbd image size could change after manually created, so after recovering the disk config we need to let other gateways to properly set the new size to LIO device. I need to propagate this info via the `config["disks"][disk_key]`, and after other gateways finishing this they need to remove the related info from the `config["disks"][disk_key]`. 

The `"recovery"` item will include the ip list of none-auth gateways.

idryomov commented:
> Actually the first 3 commits could fix the bz. The last 2 are one further optimization. I am okay to move them into a separate PR.

Is it the first 2 commits or the first 3 commits?  There are only 2 commits in this PR now...  Should "lun: try to recovery the config from the LIO" commit be re-added?

lxbsz commented:
> > Actually the first 3 commits could fix the bz. The last 2 are one further optimization. I am okay to move them into a separate PR.
> 
> Is it the first 2 commits or the first 3 commits? There are only 2 commits in this PR now... Should "lun: try to recovery the config from the LIO" commit be re-added?

I just now checked the BZ, actually the last 3 commits will fix the second bugs mentioned in the BZ.

For now I just simplified the last 3 commit and appended it here.

lxbsz commented:
@idryomov 
My test is done. Please take a look. Thanks.

idryomov commented:
Otherwise LGTM.  Any idea why the tox job is stuck `Pending`?

lxbsz commented on ceph_iscsi_config/lun.py:
```diff
@@ -574,6 +574,31 @@ def activate(self):
             if client_err:
                 raise CephiSCSIError(client_err)
 
+    def add_disk_item(self, wwn, pool_id, recovery=False):
```
Will fix it.

lxbsz commented on ceph_iscsi_config/lun.py:
```diff
@@ -574,6 +574,31 @@ def activate(self):
+                                self.config_key,
+                                disk_attr)
+
+        self.logger.debug("(LUN.allocate) registered '{}' with "
```
Sure.

lxbsz commented on ceph_iscsi_config/lun.py:
```diff
@@ -821,6 +834,11 @@ def lio_size_ok(self, rbd_object, stg_object):
         # we have the right size for the rbd - check that LIO dev size matches
         if rbd_size_ok:
+            if equal_size:
+                if stg_object.size == self.size_bytes:
+                    return True
+                else:
+                    return False
```
Will fix it.

lxbsz commented:
> Otherwise LGTM. Any idea why the tox job is stuck `Pending`?

Not sure. it seems long time ago ?

idryomov commented:
jenkins tox

idryomov commented:
@lxbsz LGTM, I'm experimenting with getting the tox tests back.
